### PR TITLE
fix the bug that terraform read nil attachment resource when the attached resource  has removed in resource CLB and CAM

### PR DESCRIPTION
**Pull request by gailwang**
1.fix doc example error of resource `tencentcloud_nat_gateway`
2.fix the bug that terraform read nil attachment resource when the attached resource of attachment resource is removed of resource CLB and CAM
3.fix the phone num of example file of CAM
4.add certificate type check in CLB

### CODE REVIEW

**Comment by gailwang** [comment]
CAM
```
 make testacc TESTARGS="-run=TestAccTencentCloudCam"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudCam -timeout 120m
# github.com/terraform-providers/terraform-provider-tencentcloud/test
test\test.go:34:4: Println call has possible formatting directive %s
?       github.com/terraform-providers/terraform-provider-tencentcloud  [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-tencentcloud/gendoc   (cached) [no tests to run]
=== RUN   TestAccTencentCloudCamGroupMembershipsDataSource_basic
--- PASS: TestAccTencentCloudCamGroupMembershipsDataSource_basic (8.36s)
=== RUN   TestAccTencentCloudCamGroupPolicyAttachmentsDataSource_basic
--- PASS: TestAccTencentCloudCamGroupPolicyAttachmentsDataSource_basic (7.29s)
=== RUN   TestAccTencentCloudCamGroupsDataSource_basic
--- PASS: TestAccTencentCloudCamGroupsDataSource_basic (3.93s)
=== RUN   TestAccTencentCloudCamPoliciesDataSource_basic
--- PASS: TestAccTencentCloudCamPoliciesDataSource_basic (4.16s)
=== RUN   TestAccTencentCloudCamRolePolicyAttachmentsDataSource_basic
--- PASS: TestAccTencentCloudCamRolePolicyAttachmentsDataSource_basic (7.21s)
=== RUN   TestAccTencentCloudCamRolesDataSource_basic
--- PASS: TestAccTencentCloudCamRolesDataSource_basic (3.74s)
=== RUN   TestAccTencentCloudCamSAMLProvidersDataSource_basic
--- PASS: TestAccTencentCloudCamSAMLProvidersDataSource_basic (3.61s)
=== RUN   TestAccTencentCloudCamUserPolicyAttachmentsDataSource_basic
--- PASS: TestAccTencentCloudCamUserPolicyAttachmentsDataSource_basic (8.01s)
=== RUN   TestAccTencentCloudCamUsersDataSource_basic
--- PASS: TestAccTencentCloudCamUsersDataSource_basic (4.27s)
=== RUN   TestAccTencentCloudCamGroupMembership_basic
--- PASS: TestAccTencentCloudCamGroupMembership_basic (12.60s)
=== RUN   TestAccTencentCloudCamGroupPolicyAttachment_basic
--- PASS: TestAccTencentCloudCamGroupPolicyAttachment_basic (7.29s)
=== RUN   TestAccTencentCloudCamGroup_basic
--- PASS: TestAccTencentCloudCamGroup_basic (4.29s)
=== RUN   TestAccTencentCloudCamPolicy_basic
--- PASS: TestAccTencentCloudCamPolicy_basic (4.33s)
=== RUN   TestAccTencentCloudCamRolePolicyAttachment_basic
--- PASS: TestAccTencentCloudCamRolePolicyAttachment_basic (7.24s)
=== RUN   TestAccTencentCloudCamRole_basic
--- PASS: TestAccTencentCloudCamRole_basic (7.30s)
=== RUN   TestAccTencentCloudCamSAMLProvider_basic
--- PASS: TestAccTencentCloudCamSAMLProvider_basic (3.77s)
=== RUN   TestAccTencentCloudCamUserPolicyAttachment_basic
--- PASS: TestAccTencentCloudCamUserPolicyAttachment_basic (7.94s)
=== RUN   TestAccTencentCloudCamUser_basic
--- PASS: TestAccTencentCloudCamUser_basic (4.60s)
=== RUN   TestAccTencentCloudCamUser_nilPassword
--- PASS: TestAccTencentCloudCamUser_nilPassword (4.17s)
=== RUN   TestAccTencentCloudCamUser_withoutKey
--- PASS: TestAccTencentCloudCamUser_withoutKey (3.87s)
PASS
ok      github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud     118.682s
?       github.com/terraform-providers/terraform-provider-tencentcloud/tencentcloud/connectivity        [no test files]

```

CLB

```
 make testacc TESTARGS="-run=TestAccTencentCloudClb"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudClb -timeout 120m
# github.com/terraform-providers/terraform-provider-tencentcloud/test
test\test.go:34:4: Println call has possible formatting directive %s
?       github.com/terraform-providers/terraform-provider-tencentcloud  [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-tencentcloud/gendoc   (cached) [no tests to run]
=== RUN   TestAccTencentCloudClbInstancesDataSource_internal
=== PAUSE TestAccTencentCloudClbInstancesDataSource_internal
=== RUN   TestAccTencentCloudClbInstancesDataSource_open
=== PAUSE TestAccTencentCloudClbInstancesDataSource_open
=== RUN   TestAccTencentCloudClbListenerRulesDataSource
=== PAUSE TestAccTencentCloudClbListenerRulesDataSource
=== RUN   TestAccTencentCloudClbListenersDataSource
=== PAUSE TestAccTencentCloudClbListenersDataSource
=== RUN   TestAccTencentCloudClbRedirectionsDataSource
=== PAUSE TestAccTencentCloudClbRedirectionsDataSource
=== RUN   TestAccTencentCloudClbServerAttachment_tcp
--- PASS: TestAccTencentCloudClbServerAttachment_tcp (42.18s)
=== RUN   TestAccTencentCloudClbServerAttachment_http
--- PASS: TestAccTencentCloudClbServerAttachment_http (44.57s)
=== RUN   TestAccTencentCloudClbInstance_basic
=== PAUSE TestAccTencentCloudClbInstance_basic
=== RUN   TestAccTencentCloudClbInstance_open
=== PAUSE TestAccTencentCloudClbInstance_open
=== RUN   TestAccTencentCloudClbInstance_internal
=== PAUSE TestAccTencentCloudClbInstance_internal
=== RUN   TestAccTencentCloudClbListenerRule_basic
=== PAUSE TestAccTencentCloudClbListenerRule_basic
=== RUN   TestAccTencentCloudClbListenerRule_full
=== PAUSE TestAccTencentCloudClbListenerRule_full
=== RUN   TestAccTencentCloudClbListener_basic
=== PAUSE TestAccTencentCloudClbListener_basic
=== RUN   TestAccTencentCloudClbListener_tcp
=== PAUSE TestAccTencentCloudClbListener_tcp
=== RUN   TestAccTencentCloudClbListener_https
=== PAUSE TestAccTencentCloudClbListener_https
=== RUN   TestAccTencentCloudClbListener_tcpssl
=== PAUSE TestAccTencentCloudClbListener_tcpssl
=== RUN   TestAccTencentCloudClbRedirection_basic
--- PASS: TestAccTencentCloudClbRedirection_basic (63.92s)
=== RUN   TestAccTencentCloudClbRedirection_auto
--- PASS: TestAccTencentCloudClbRedirection_auto (64.32s)
=== CONT  TestAccTencentCloudClbInstancesDataSource_internal
=== CONT  TestAccTencentCloudClbListenerRule_basic
=== CONT  TestAccTencentCloudClbRedirectionsDataSource
=== CONT  TestAccTencentCloudClbInstance_open
=== CONT  TestAccTencentCloudClbListenerRulesDataSource
=== CONT  TestAccTencentCloudClbListener_https
=== CONT  TestAccTencentCloudClbInstancesDataSource_open
=== CONT  TestAccTencentCloudClbListener_basic
--- PASS: TestAccTencentCloudClbInstancesDataSource_open (41.33s)
=== CONT  TestAccTencentCloudClbInstance_internal
--- PASS: TestAccTencentCloudClbInstancesDataSource_internal (43.82s)
=== CONT  TestAccTencentCloudClbListenerRule_full
--- PASS: TestAccTencentCloudClbInstance_open (116.79s)
=== CONT  TestAccTencentCloudClbListenersDataSource
--- PASS: TestAccTencentCloudClbListener_basic (129.21s)
=== CONT  TestAccTencentCloudClbInstance_basic
--- PASS: TestAccTencentCloudClbInstance_internal (139.36s)
=== CONT  TestAccTencentCloudClbListener_tcp
--- PASS: TestAccTencentCloudClbInstance_basic (52.08s)
=== CONT  TestAccTencentCloudClbListener_tcpssl
--- PASS: TestAccTencentCloudClbListenerRule_basic (183.42s)
--- PASS: TestAccTencentCloudClbListenerRulesDataSource (185.67s)
--- PASS: TestAccTencentCloudClbListener_https (192.84s)
--- PASS: TestAccTencentCloudClbListenersDataSource (106.29s)
--- PASS: TestAccTencentCloudClbRedirectionsDataSource (252.23s)
--- PASS: TestAccTencentCloudClbListenerRule_full (237.49s)
--- PASS: TestAccTencentCloudClbListener_tcp (107.80s)
--- PASS: TestAccTencentCloudClbListener_tcpssl (109.32s)
PASS
ok 
```

**Comment by gailwang** [comment]
CAM 控制台删除group 再次apply 测试
``` 
terraform apply
tencentcloud_cam_group.example: Refreshing state... [id=99728]
tencentcloud_cam_policy.example: Refreshing state... [id=29375879]
tencentcloud_cam_role.example: Refreshing state... [id=4611686018427989452]
tencentcloud_cam_user.example: Refreshing state... [id=example]
tencentcloud_cam_role_policy_attachment.example: Refreshing state... [id=4611686018427989452#29375879]
tencentcloud_cam_user_policy_attachment.example: Refreshing state... [id=example#29375879]
tencentcloud_cam_group_policy_attachment.example: Refreshing state... [id=99728#29375879]
tencentcloud_cam_group_membership.example: Refreshing state... [id=99728]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # tencentcloud_cam_group.example will be created
  + resource "tencentcloud_cam_group" "example" {
      + create_time = (known after apply)
      + id          = (known after apply)
      + name        = "example"
      + remark      = "example"
    }

  # tencentcloud_cam_group_membership.example will be created
  + resource "tencentcloud_cam_group_membership" "example" {
      + group_id = (known after apply)
      + id       = (known after apply)
      + user_ids = [
          + "example",
        ]
    }

  # tencentcloud_cam_group_policy_attachment.example will be created
  + resource "tencentcloud_cam_group_policy_attachment" "example" {
      + create_mode = (known after apply)
      + create_time = (known after apply)
      + group_id    = (known after apply)
      + id          = (known after apply)
      + policy_id   = "29375879"
      + policy_name = (known after apply)
      + policy_type = (known after apply)
    }

Plan: 3 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

tencentcloud_cam_group.example: Creating...
tencentcloud_cam_group.example: Creation complete after 3s [id=99729]
tencentcloud_cam_group_policy_attachment.example: Creating...
tencentcloud_cam_group_membership.example: Creating...
tencentcloud_cam_group_policy_attachment.example: Creation complete after 3s [id=99729#29375879]
tencentcloud_cam_group_membership.example: Creation complete after 3s [id=99729]

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.
```

CAM 控制台删除policy 再次apply 测试
```
 terraform apply
tencentcloud_cam_policy.example: Refreshing state... [id=29375879]
tencentcloud_cam_group.example: Refreshing state... [id=99729]
tencentcloud_cam_user.example: Refreshing state... [id=example]
tencentcloud_cam_role.example: Refreshing state... [id=4611686018427989452]
tencentcloud_cam_user_policy_attachment.example: Refreshing state... [id=example#29375879]
tencentcloud_cam_role_policy_attachment.example: Refreshing state... [id=4611686018427989452#29375879]
tencentcloud_cam_group_membership.example: Refreshing state... [id=99729]
tencentcloud_cam_group_policy_attachment.example: Refreshing state... [id=99729#29375879]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # tencentcloud_cam_group_policy_attachment.example will be created
  + resource "tencentcloud_cam_group_policy_attachment" "example" {
      + create_mode = (known after apply)
      + create_time = (known after apply)
      + group_id    = "99729"
      + id          = (known after apply)
      + policy_id   = (known after apply)
      + policy_name = (known after apply)
      + policy_type = (known after apply)
    }

  # tencentcloud_cam_policy.example will be created
  + resource "tencentcloud_cam_policy" "example" {
      + create_time = (known after apply)
      + document    = jsonencode(
            {
              + statement = [
                  + {
                      + action   = [
                          + "name/sts:AssumeRole",
                        ]
                      + effect   = "allow"
                      + resource = [
                          + "*",
                        ]
                    },
                  + {
                      + action   = [
                          + "name/cos:PutObject",
                        ]
                      + effect   = "allow"
                      + resource = [
                          + "*",
                        ]
                    },
                ]
              + version   = "2.0"
            }
        )
      + id          = (known after apply)
      + name        = "example"
      + type        = (known after apply)
      + update_time = (known after apply)
    }

  # tencentcloud_cam_role_policy_attachment.example will be created
  + resource "tencentcloud_cam_role_policy_attachment" "example" {
      + create_mode = (known after apply)
      + create_time = (known after apply)
      + id          = (known after apply)
      + policy_id   = (known after apply)
      + policy_name = (known after apply)
      + policy_type = (known after apply)
      + role_id     = "4611686018427989452"
    }

  # tencentcloud_cam_user_policy_attachment.example will be created
  + resource "tencentcloud_cam_user_policy_attachment" "example" {
      + create_mode = (known after apply)
      + create_time = (known after apply)
      + id          = (known after apply)
      + policy_id   = (known after apply)
      + policy_name = (known after apply)
      + policy_type = (known after apply)
      + user_id     = "example"
    }

Plan: 4 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

tencentcloud_cam_policy.example: Creating...
tencentcloud_cam_policy.example: Creation complete after 3s [id=29400104]
tencentcloud_cam_role_policy_attachment.example: Creating...
tencentcloud_cam_user_policy_attachment.example: Creating...
tencentcloud_cam_group_policy_attachment.example: Creating...
tencentcloud_cam_role_policy_attachment.example: Creation complete after 4s [id=4611686018427989452#29400104]
tencentcloud_cam_group_policy_attachment.example: Creation complete after 4s [id=99729#29400104]
tencentcloud_cam_user_policy_attachment.example: Creation complete after 4s [id=example#29400104]

Apply complete! Resources: 4 added, 0 changed, 0 destroyed.

```

CAM删除role再次apply测试

```
terraform apply
tencentcloud_cam_policy.example: Refreshing state... [id=29400104]
tencentcloud_cam_role.example: Refreshing state... [id=4611686018427989452]
tencentcloud_cam_user.example: Refreshing state... [id=example]
tencentcloud_cam_group.example: Refreshing state... [id=99729]
tencentcloud_cam_user_policy_attachment.example: Refreshing state... [id=example#29400104]
tencentcloud_cam_role_policy_attachment.example: Refreshing state... [id=4611686018427989452#29400104]
tencentcloud_cam_group_membership.example: Refreshing state... [id=99729]
tencentcloud_cam_group_policy_attachment.example: Refreshing state... [id=99729#29400104]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # tencentcloud_cam_role.example will be created
  + resource "tencentcloud_cam_role" "example" {
      + console_login = true
      + create_time   = (known after apply)
      + description   = "test"
      + document      = jsonencode(
            {
              + statement = [
                  + {
                      + action    = [
                          + "name/sts:AssumeRole",
                        ]
                      + effect    = "allow"
                      + principal = {
                          + qcs = [
                              + "qcs::cam::uin/100009461222:uin/100009461222",
                            ]
                        }
                    },
                ]
              + version   = "2.0"
            }
        )
      + id            = (known after apply)
      + name          = "example"
      + update_time   = (known after apply)
    }

  # tencentcloud_cam_role_policy_attachment.example will be created
  + resource "tencentcloud_cam_role_policy_attachment" "example" {
      + create_mode = (known after apply)
      + create_time = (known after apply)
      + id          = (known after apply)
      + policy_id   = "29400104"
      + policy_name = (known after apply)
      + policy_type = (known after apply)
      + role_id     = (known after apply)
    }

Plan: 2 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

tencentcloud_cam_role.example: Creating...
tencentcloud_cam_role.example: Creation complete after 4s [id=4611686018427989452]
tencentcloud_cam_role_policy_attachment.example: Creating...
tencentcloud_cam_role_policy_attachment.example: Creation complete after 3s [id=4611686018427989452#29400104]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```
CAM 删除user测试
```
 terraform apply
tencentcloud_cam_policy.example: Refreshing state... [id=29400104]
tencentcloud_cam_group.example: Refreshing state... [id=99729]
tencentcloud_cam_user.example: Refreshing state... [id=example]
tencentcloud_cam_role.example: Refreshing state... [id=4611686018427989452]
tencentcloud_cam_user_policy_attachment.example: Refreshing state... [id=example#29400104]
tencentcloud_cam_role_policy_attachment.example: Refreshing state... [id=4611686018427989452#29400104]
tencentcloud_cam_group_membership.example: Refreshing state... [id=99729]
tencentcloud_cam_group_policy_attachment.example: Refreshing state... [id=99729#29400104]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # tencentcloud_cam_group_membership.example will be created
  + resource "tencentcloud_cam_group_membership" "example" {
      + group_id = "99729"
      + id       = (known after apply)
      + user_ids = (known after apply)
    }

  # tencentcloud_cam_user.example will be created
  + resource "tencentcloud_cam_user" "example" {
      + console_login       = true
      + country_code        = "86"
      + email               = "1234@qq.com"
      + id                  = (known after apply)
      + name                = "example"
      + need_reset_password = true
      + password            = (sensitive value)
      + phone_num           = "13631555963"
      + remark              = "example"
      + secret_id           = (sensitive value)
      + secret_key          = (sensitive value)
      + uid                 = (known after apply)
      + uin                 = (known after apply)
      + use_api             = true
    }

  # tencentcloud_cam_user_policy_attachment.example will be created
  + resource "tencentcloud_cam_user_policy_attachment" "example" {
      + create_mode = (known after apply)
      + create_time = (known after apply)
      + id          = (known after apply)
      + policy_id   = "29400104"
      + policy_name = (known after apply)
      + policy_type = (known after apply)
      + user_id     = (known after apply)
    }

Plan: 3 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

tencentcloud_cam_user.example: Creating...
tencentcloud_cam_user.example: Creation complete after 3s [id=example]
tencentcloud_cam_user_policy_attachment.example: Creating...
tencentcloud_cam_group_membership.example: Creating...
tencentcloud_cam_group_membership.example: Creation complete after 4s [id=99729]
tencentcloud_cam_user_policy_attachment.example: Creation complete after 4s [id=example#29400104]

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.
```

File: tencentcloud/data_source_tc_clb_listener_rules.go [hunk 1]
```diff
@@ -242,7 +242,9 @@ func dataSourceTencentCloudClbListenerRulesRead(d *schema.ResourceData, meta int
 		if rule.Certificate != nil {
 			mapping["certificate_ssl_mode"] = *rule.Certificate.SSLMode
 			mapping["certificate_id"] = *rule.Certificate.CertId
-			mapping["certificate_ca_id"] = *rule.Certificate.CertCaId
+			if rule.Certificate.CertCaId != nil {
```
Comment (Sherlock-Holo) [inline tencentcloud/data_source_tc_clb_listener_rules.go]: 可以直接
`mapping["certificate_ca_id"] = rule.Certificate.CertCaId`

File: tencentcloud/data_source_tc_clb_listeners.go [hunk 2]
```diff
@@ -221,7 +221,9 @@ func dataSourceTencentCloudClbListenersRead(d *schema.ResourceData, meta interfa
 		if listener.Certificate != nil {
 			mapping["certificate_ssl_mode"] = *listener.Certificate.SSLMode
 			mapping["certificate_id"] = *listener.Certificate.CertId
-			mapping["certificate_ca_id"] = *listener.Certificate.CertCaId
+			if listener.Certificate.CertCaId != nil {
```
Comment (Sherlock-Holo) [inline tencentcloud/data_source_tc_clb_listeners.go]: 同上

File: tencentcloud/resource_tc_clb_attachment.go [hunk 3]
```diff
@@ -175,9 +175,19 @@ func resourceTencentCloudClbServerAttachmentDelete(d *schema.ResourceData, meta
 	listenerId := items[1]
 	clbId := items[2]
 
+	//firstly see if listener and location exists
+	request := clb.NewDeregisterTargetsRequest()
+	request.ListenerId = stringToPointer(listenerId)
```
Comment (oliverpei) [inline tencentcloud/resource_tc_clb_attachment.go]: 可以直接写 request.ListenerId = &listenerId, 不用调用函数

**Pull request merged**